gateway-js/src/index.ts

```diff
@@ -425,135 +416,143 @@ export class ApolloGateway implements GraphQLService {
     };
   }
 
-  // Synchronously load a statically configured schema, update class instance's
-  // schema and query planner.
-  private loadStatic(config: StaticGatewayConfig) {
-    let schema: GraphQLSchema;
-    let supergraphSdl: string;
-    try {
-      ({ schema, supergraphSdl } = isLocalConfig(config)
-        ? this.createSchemaFromServiceList(config.localServiceList)
-        : this.createSchemaFromSupergraphSdl(config.supergraphSdl));
-      // TODO(trevor): #580 redundant parse
-      this.parsedSupergraphSdl = parse(supergraphSdl);
-      this.supergraphSdl = supergraphSdl;
-      this.updateWithSchemaAndNotify(schema, supergraphSdl, true);
-    } catch (e) {
-      this.state = { phase: 'failed to load' };
-      throw e;
-    }
-    this.state = { phase: 'loaded' };
+  private getUplinkEndpoints(config: ManagedGatewayConfig) {
+    // 1. If config is set to a `string`, use it
+    // 2. If the env var is set, use that
+    // 3. If config is `undefined`, use the default uplink URLs
+    const rawEndpointsString =
+      process.env.APOLLO_SCHEMA_CONFIG_DELIVERY_ENDPOINT;
+    const envEndpoints = rawEndpointsString?.split(',') ?? null;
+    return config.uplinkEndpoints ??
+      (config.schemaConfigDeliveryEndpoint
+        ? [config.schemaConfigDeliveryEndpoint]
+        : null) ??
+      envEndpoints ?? [
+        'https://uplink.api.apollographql.com/',
+        'https://aws.uplink.api.apollographql.com/',
+      ];
+  }
+
+  private getIdForSupergraphSdl(supergraphSdl: string) {
+    return createHash('sha256').update(supergraphSdl).digest('hex');
   }
 
-  // Asynchronously load a dynamically configured schema. `this.updateSchema`
-  // is responsible for updating the class instance's schema and query planner.
-  private async loadDynamic(unrefTimer: boolean) {
+  private async initializeSupergraphManager<T extends SupergraphManager>(
+    supergraphSdlFetcher: T,
+  ) {
     try {
-      await this.updateSchema();
+      const result = await supergraphSdlFetcher.initialize({
+        update: this.externalSupergraphUpdateCallback.bind(this),
+        healthCheck: this.externalSubgraphHealthCheckCallback.bind(this),
+        getDataSource: this.externalGetDataSourceCallback.bind(this),
+      });
+      if (!result?.supergraphSdl) {
+        throw new Error(
+          'Provided `supergraphSdl` function did not return an object containing a `supergraphSdl` property',
+        );
+      }
+      if (result?.cleanup) {
```

Comment:
Potentially user provided, maybe make sure this is typeof function and warn if not

Reply:
8df89e77